### PR TITLE
UCP/WIREUP: added multi-lane initialization

### PR DESCRIPTION
**Pull request by hoopoepg**
- added initialization of multi-lane rndv-get transport

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5297/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@brminich any ideas?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5305/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by hoopoepg** [comment]
bot:bgate:retest

File: src/ucp/wireup/select.c [hunk 1]
```diff
@@ -769,23 +780,33 @@ static ucs_status_t ucp_wireup_add_am_lane(ucp_ep_h ep, const ucp_ep_params_t *p
...
+    ucp_lane_index_t max_lanes;
+    ucp_lane_index_t i;
 
     if (!(ucp_ep_get_context_features(ep) & UCP_FEATURE_TAG)) {
         return UCS_OK;
     }
 
+    if ((params->field_mask & UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE) &&
+        (params->err_mode == UCP_ERR_HANDLING_MODE_PEER)) {
+        ucs_trace_func("rendezvous: disabled due to error mode");
```
Comment (brminich) [inline src/ucp/wireup/select.c]: minor: we typically use `ucs_trace` for wireup

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -801,16 +822,52 @@ static ucs_status_t ucp_wireup_add_rndv_lane(ucp_ep_h ep,
...
+                                             tl_bitmap, -1, 0, &rsc_index, &addr_index, &score);
+        if ((status == UCS_OK) &&
+            /* a temporary workaround to prevent the ugni uct from using rndv */
+            (strstr(ep->worker->context->tl_rscs[rsc_index].tl_rsc.tl_name, "ugni") == NULL)) {
+             ucp_wireup_add_lane_desc(lane_descs, num_lanes_p, rsc_index, addr_index,
+                                      address_list[addr_index].md_index, score,
+                                      UCP_WIREUP_LANE_USAGE_RNDV, 0);
+             i++;
+        } else if (status != UCS_OK) {
+            break;
```
Comment (brminich) [inline src/ucp/wireup/select.c]: you break here, but will return UCS_OK in any case. Is it expected?

Comment (hoopoepg) [reply src/ucp/wireup/select.c]: yes, it breaks when no more lanes are available

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -401,6 +406,12 @@ static int ucp_wireup_compare_lane_amo_score(const void *elem1, const void *elem
     return UCP_WIREUP_COMPARE_SCORE(elem1, elem2, arg, amo);
 }
 
+static int ucp_wireup_compare_lane_rndv_score(const void *elem1, const void *elem2,
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: the same as all other inlined funcs here

File: src/ucp/wireup/select.c (hunk 1 above)
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: will fix after tests are complete

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5309/ for details (Mellanox internal link).

File: src/ucp/wireup/select.c (hunk 1 above)
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: fixed

File: src/ucp/wireup/select.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucp/wireup/select.c]: I will add separate PR to unify INLINE macro

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5315/ for details (Mellanox internal link).

File: src/ucp/wireup/select.c [hunk 4]
```diff
@@ -801,16 +822,52 @@ static ucs_status_t ucp_wireup_add_rndv_lane(ucp_ep_h ep,
...
+            (strstr(ep->worker->context->tl_rscs[rsc_index].tl_rsc.tl_name, "ugni") == NULL)) {
+             ucp_wireup_add_lane_desc(lane_descs, num_lanes_p, rsc_index, addr_index,
+                                      address_list[addr_index].md_index, score,
+                                      UCP_WIREUP_LANE_USAGE_RNDV, 0);
+             i++;
+        } else if (status != UCS_OK) {
+            break;
+        }
+
+        if (ep->worker->context->tl_rscs[rsc_index].tl_rsc.dev_type == UCT_DEVICE_TYPE_SHM) {
```
Comment (yosefe) [inline src/ucp/wireup/select.c]: don't we also need to check that same device is not used for more than one rndv lane?

Comment (hoopoepg) [reply src/ucp/wireup/select.c]: every found rsc_index is excluded from next processing

Comment (yosefe) [reply src/ucp/wireup/select.c]: true, but if we have 2 transports on same device (e.g dc ad rc) they would be different rsc index

Comment (hoopoepg) [reply src/ucp/wireup/select.c]: ok, fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5335/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5342/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
👍

**Comment by yosefe** [comment]
will need to be reworked to select multiple RMA lanes

**Pull request closed without merging**